control-plane/connect-inject/controllers/pod/pod_controller.go

```diff
@@ -210,33 +233,238 @@ func (r *Controller) writeWorkload(ctx context.Context, pod corev1.Pod) error {
+		return nil
+	}
+
+	pc := &pbmesh.ProxyConfiguration{
+		Workloads: &pbcatalog.WorkloadSelector{
+			Names: []string{pod.GetName()},
+		},
+		DynamicConfig: &pbmesh.DynamicConfig{
+			Mode:         mode,
+			ExposeConfig: exposeConfig,
+		},
+		BootstrapConfig: bootstrapConfig,
+	}
+
+	// TODO(dans): replace with common.ToProtoAny when available
+	data, err := anypb.New(pc)
+	if err != nil {
+		return fmt.Errorf("could not serialize workload: %w", err)
 	}
 
 	req := &pbresource.WriteRequest{
 		Resource: &pbresource.Resource{
-			Id:       getWorkloadID(pod.GetName(), r.getConsulNamespace(pod.Namespace), r.getPartition()),
-			Metadata: meta,
-			Data:     proto,
+			Id:       getProxyConfigurationID(pod.GetName(), r.getConsulNamespace(pod.Namespace), r.getPartition()),
+			Metadata: metaFromPod(pod),
+			Data:     data,
 		},
 	}
 	_, err = r.ResourceClient.Write(ctx, req)
 	return err
 }
 
-//func (r *Controller) writeHealthStatus(pod corev1.Pod) error {
-//	return nil
-//}
+func (r *Controller) getTproxyMode(ctx context.Context, pod corev1.Pod) (pbmesh.ProxyMode, error) {
+	// A user can enable/disable tproxy for an entire namespace.
+	var ns corev1.Namespace
+	err := r.Client.Get(ctx, types.NamespacedName{Name: pod.GetNamespace()}, &ns)
+	if err != nil {
+		return pbmesh.ProxyMode_PROXY_MODE_DEFAULT, fmt.Errorf("could not get namespace info for %s: %w", pod.GetNamespace(), err)
+	}
+
+	tproxyEnabled, err := common.TransparentProxyEnabled(ns, pod, r.EnableTransparentProxy)
+	if err != nil {
+		return pbmesh.ProxyMode_PROXY_MODE_DEFAULT, fmt.Errorf("could not determine if transparent proxy is enabled: %w", err)
+	}
+
+	if tproxyEnabled {
+		return pbmesh.ProxyMode_PROXY_MODE_TRANSPARENT, nil
+	}
+	return pbmesh.ProxyMode_PROXY_MODE_DEFAULT, nil
+
+}
+
+func (r *Controller) getExposeConfig(pod corev1.Pod) (*pbmesh.ExposeConfig, error) {
+
+	// Expose k8s probes as Envoy listeners if needed.
+	overwriteProbes, err := common.ShouldOverwriteProbes(pod, r.TProxyOverwriteProbes)
+	if err != nil {
+		return nil, fmt.Errorf("could not determine if probes should overwrite: %w", err)
+	}
+
+	if !overwriteProbes {
+		return nil, nil
+	}
 
-// TODO(dans): delete ACL token for workload
+	var originalPod corev1.Pod
+	err = json.Unmarshal([]byte(pod.Annotations[constants.AnnotationOriginalPod]), &originalPod)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get original pod spec: %w", err)
+	}
+
+	exposeConfig := &pbmesh.ExposeConfig{}
+	for _, mutatedContainer := range pod.Spec.Containers {
+		for _, originalContainer := range originalPod.Spec.Containers {
+			if originalContainer.Name == mutatedContainer.Name {
+				paths, err := getContainerExposePaths(originalPod, originalContainer, mutatedContainer)
+				if err != nil {
+					return nil, fmt.Errorf("error getting container expose path for %s: %w", originalContainer.Name, err)
+				}
+
+				exposeConfig.ExposePaths = append(exposeConfig.ExposePaths, paths...)
+			}
+		}
+	}
```

Comment:
Just to call out, this looks like a straight port of Endpoints V1, so I assumed it was meant to be 1-1 and didn't read closely. If this needs more thorough review happy to try myself or ask for a hand.

Reply:
Yep, this is verbatim V1 Endpoints code, shifted around a bit to make it easier to read.